### PR TITLE
Core: detect incorrectly nested elements

### PR DESCRIPTION
**Pull request by dmethvin**
Basic approach: If our old `jQuery.htmlPrefilter` would have changed the string _and_ the resulting actual HTML string is different, give a warning.

### CODE REVIEW

File: src/core.js [hunk 1]
```diff
@@ -3,13 +3,28 @@ var oldInit = jQuery.fn.init,
...
 	rattrHashGlob = /\[(\s*[-\w]+\s*)([~|^$*]?=)\s*([-\w#]*?#[-\w#]*)\s*\]/g,
+	rxhtmlTag = /<(?!area|br|col|embed|hr|img|input|link|meta|param)(([a-z][^\/\0>\x20\t\r\n\f]*)[^>]*)\/>/gi,
+	rsingleTag = ( /^<([a-z][^\/\0>:\x20\t\r\n\f]*)[\x20\t\r\n\f]*\/?>(?:<\/\1>|)$/i ),
 
 	// Support: Android <=4.0 only
 	// Make sure we trim BOM and NBSP
 	rtrim = /^[\s\uFEFF\xA0]+|[\s\uFEFF\xA0]+$/g;
 
+var makeMarkup = function( html ) {
+	var createDoc = window.document.implementation.createHTMLDocument;
```
Comment (dmethvin) [inline src/core.js:14]: Not completely sure this guard is needed yet.

**Comment by dmethvin** [comment]
@mgol  This seems to work pretty well. Thoughts?

File: src/core.js [hunk 2]
```diff
@@ -18,6 +33,15 @@ jQuery.fn.init = function( arg1 ) {
 		args[ 0 ] = [];
 	}
 
+	if ( typeof arg1 === "string" && arg1.trim().charAt( 0 ) === "<" ) {
+		if ( !rsingleTag.test( arg1 ) ) {
+			changed = arg1.replace( rxhtmlTag, "<$1></$2>" );
+			if ( changed !== arg1 && makeMarkup( arg1 ) !== makeMarkup( changed ) ) {
+				migrateWarn( "HTML tags must be properly nested and closed: " + arg1 );
```
Comment (mgol) [inline src/core.js:40]: I was thinking about doing it by overwriting `jQuery.htmlPrefilter` instead; then we would miss no cases & the logic would be simpler. Of course, when someone overwrites it on their end, that would remove the warnings but maybe that's acceptable? We could make it more explicit in the [3.5 upgrade guide](https://jquery.com/upgrade-guide/3.5/) that if Migrate is used, it's best to avoid manually overwriting the method?

Doing it in `jQuery.fn.init` probably catches the majority of usage but I suspect it leaves a significant portion out. Consider e.g.:
```js
jQuery( "#test" ).append( "<div/><span/>" );
```
We'd get no warning for that.

**Comment by dmethvin** [comment]
Using `.htmlPrefilter()` seems like a good idea, I'll switch to that and add a few more tests.

File: src/core.js [hunk 3]
```diff
@@ -22,6 +36,19 @@ jQuery.fn.init = function( arg1 ) {
 };
 jQuery.fn.init.prototype = jQuery.fn;
 
+jQuery.htmlPrefilter = function( html ) {
+	var changed;
+	if ( typeof html === "string" && html.trim().charAt( 0 ) === "<" ) {
```
Comment (dmethvin) [inline src/core.js:41]: Not sure if this guard is needed but it shouldn't hurt.

File: test/core.js [hunk 4]
```diff
@@ -1,84 +1,122 @@
 
-QUnit.module( "core" );
```
Comment (dmethvin) [inline test/core.js:2]: I'm happier with this lower-space version but not sure why it happened. Did we change some config file?

File: test/core.js [hunk 5]
```diff
@@ -1,84 +1,122 @@
...
+		jQuery("<blockquote><p class='bad' /><span/></blockquote>");
+		jQuery("<div />").append("<div data-borked='y'><span/> <p/></div>");
+		jQuery("<script /><div class=afterscript/>")
+	});
+
+	expectNoWarning(assert, "Script doesn't run in processed strings", function () {
+		jQuery("<script>alert('YOU SHOULD NOT SEE THIS ALERT')</script>");
+		jQuery("<div>").append("<script src='YOU-SHOULD-NOT-SEE-THIS.SCRIPT'></script>");
+		jQuery("<script>alert('YOU SHOULD NOT SEE THIS CONSOLE')</script>");
+		// jQuery core code causes this to execute, and also it's async
```
Comment (dmethvin) [inline test/core.js:87]: I wonder if we should use this approach for jQuery 4.0 to avoid running script in the initial processing.

**Comment by dmethvin** [comment]
@mgol Take a look when you get a chance and give me your thoughts.

File: src/core.js (hunk 3 above)
Comment (mgol) [reply src/core.js:41]: The original `jQuery.htmlPrefilter` in jQuery 3.4.1 & older is just:
```js
jQuery.htmlPrefilter = function( html ) {
	return html.replace( rxhtmlTag, "<$1></$2>" );
};
```
so it has no guard. Therefore, I don't think we need it here either.

File: test/core.js (hunk 4 above)
Comment (mgol) [reply test/core.js:2]: I submitted a PR that improves our ESLint setup and should resolve the issue you had here: #339. Let's get it merged first so that we don't go back & forth on code style here.

File: test/core.js (hunk 5 above)
Comment (mgol) [reply test/core.js:87]: This follows the native behavior, I wouldn't want to fight it too much.

File: test/core.js [hunk 6]
```diff
@@ -1,84 +1,122 @@
...
-				"<a href='space#junk'>test</a>" +
-				"<link rel='good#stuff' />" +
-				"<p class='space #junk'>" +
-					"<a href='#some-anchor'>anchor2</a>" +
-					"<input value='[strange*=#stuff]' />" +
-					"<a href='#' data-id='#junk'>anchor</a>" +
-				"</p>" +
-			"</div>" ).appendTo( "#qunit-fixture" ),
+		"<div>" +
+		"<div data-selector='a[href=#main]'></div>" +
```
Comment (mgol) [inline test/core.js:111]: The previous markup was more readable due to indentation. I'd keep the old styling or add indentation in the string itself, the new way is hard to read.

IMO it'd be good to revert all unrelated style changes from this PR, merge #339 & then only add new tests. We can always improve the styling in a separate PR later when we know ESLint actually checks it.

File: test/core.js (hunk 4 above)
Comment (mgol) [reply test/core.js:2]: #339 has landed so you can rebase now.

File: test/core.js [hunk 7]
```diff
@@ -60,12 +60,48 @@ QUnit.test( "jQuery( '#' )", function( assert ) {
 	assert.expect( 2 );
 
 	expectWarning( assert, "Selector, through the jQuery constructor, nothing but hash",
-			function() {
```
Comment (dmethvin) [inline test/core.js:63]: There are some spurious diffs here from line lengths exceeding 100 chars that I had to fix.

File: src/core.js [hunk 8]
```diff
@@ -25,6 +31,14 @@ jQuery.fn.init = function( arg1 ) {
 };
 jQuery.fn.init.prototype = jQuery.fn;
 
+jQuery.htmlPrefilter = function( html ) {
+	var changed = html.replace( rxhtmlTag, "<$1></$2>" );
+	if ( changed !== html && makeMarkup( html ) !== makeMarkup( changed ) ) {
+		migrateWarn( "HTML tags must be properly nested and closed: " + html );
+	}
+	return html;
+};
```
Comment (mgol) [inline src/core.js:40]: This will make `jQuery.htmlPrefilter` in jQuery <3.5 to behave like an identity function which we should avoid IMO. I'd keep the warning mechanism as is as it will prepare users for a 3.5 upgrade but instead of returning `html` I'd return the result of applying the original cached `htmlPrefilter`:
```suggestion
var origHtmlPrefilter = jQuery.htmlPrefilter;
jQuery.htmlPrefilter = function( html ) {
	var changed = html.replace( rxhtmlTag, "<$1></$2>" );
	if ( changed !== html && makeMarkup( html ) !== makeMarkup( changed ) ) {
		migrateWarn( "HTML tags must be properly nested and closed: " + html );
	}
	return origHtmlPrefilter( html );
};
```

File: test/core.js [hunk 9]
```diff
@@ -60,12 +60,48 @@ QUnit.test( "jQuery( '#' )", function( assert ) {
...
+		jQuery( "<blockquote><p class='bad' /><span/></blockquote>" );
+		jQuery( "<div />" ).append( "<div data-borked='y'><span/> <p/></div>" );
+		jQuery( "<script /><div class=afterscript/>" );
+	} );
+
+	expectNoWarning( assert, "Script doesn't run in processed strings", function() {
+		jQuery( "<script>alert( 'YOU SHOULD NOT SEE THIS ALERT' )</script>" );
+		jQuery( "<div>" ).append( "<script src='YOU-SHOULD-NOT-SEE-THIS.SCRIPT'></script>" );
+		jQuery( "<script>alert( 'YOU SHOULD NOT SEE THIS CONSOLE' )</script>" );
+	} );
```
Comment (mgol) [inline test/core.js:88]: This doesn't hurt but out of curiosity: is this related to the tag self-closing behavior?

File: src/core.js [hunk 10]
```diff
@@ -25,6 +31,14 @@ jQuery.fn.init = function( arg1 ) {
 };
 jQuery.fn.init.prototype = jQuery.fn;
 
+jQuery.htmlPrefilter = function( html ) {
```
Comment (mgol) [inline src/core.js:34]: I think this code should lie in `src/manipulation.js` as it was in jQuery. In that case you won't need a separate `rxhtmlTag` definition as it's already defined there.

File: test/core.js (hunk 9 above)
Comment (dmethvin) [reply test/core.js:88]: Mainly I just wanted to make sure we weren't introducing any script issues with this.

File: src/core.js (hunk 10 above)
Comment (dmethvin) [reply src/core.js:35]: I'll move it. I'd actually forgotten we had it there.

File: src/manipulation.js [hunk 11]
```diff
@@ -1,7 +1,21 @@
...
+		return doc.body && doc.body.innerHTML;
+	};
 
 jQuery.UNSAFE_restoreLegacyHtmlPrefilter = function() {
 	jQuery.htmlPrefilter = function( html ) {
 		return html.replace( rxhtmlTag, "<$1></$2>" );
 	};
 };
+
+jQuery.htmlPrefilter = function( html ) {
```
Comment (dmethvin) [inline src/manipulation.js:15]: Restoring the legacy filter above will clobber this warning, but it's already marked `UNSAFE_` so we can do just about anything we want including format the drive.

Comment (mgol) [reply src/manipulation.js:15]: That's true. That said, the part below with the `makeMarkup` comparison & the printing the warning could be extracted to a separate function which we could then use both here & in the `htmlPrefilter` version that `UNSAFE_restoreLegacyHtmlPrefilter` restores. What do you think?

Comment (dmethvin) [reply src/manipulation.js:15]: I wonder if someone who uses the UNSAFE_ call actually wants the warning? Probably not.

Comment (mgol) [reply src/manipulation.js:15]: @dmethvin I think many of them do! Some people may not care (but if they don't care about Migrate warnings they may just mute them all) but I imagine many users of this API won't be people _not_ wanting to adjust but ones that want to perform the jQuery update but who don't have time (or possibilities - they may rely on external code that needs to get updated) to deal with that breaking change at the moment. They will often still want to fix their jQuery usage & remove the `UNSAFE_` call. Migrate warnings would help them a lot here!

Comment (mgol) [reply src/manipulation.js:15]: I think people that want this warning silenced without silencing the others & who want the old behavior restored can follow the non-Migrate based advice from our [3.5 upgrade guide](https://jquery.com/upgrade-guide/3.5/).

But I'd expect many users of this `UNSAFE_` API to want to be warned so that they can fix violations & get rid of it as it essentially undoes a security fix.

Comment (dmethvin) [reply src/manipulation.js:22]: You're right! I could see someone restoring the filter temporarily and letting things run to see what warnings they got. Without that, their code would be broken and they might not be able to test things fully to fix them. I've added a commit, see what you think.
